File: src/main/java/com/kenvix/utils/log/LoggingOutputStream.java
//--------------------------------------------------
// Class LoggingOutputStream
//--------------------------------------------------
// Written by Kenvix <i@kenvix.com>
//--------------------------------------------------

package com.kenvix.utils.log;

import org.jetbrains.annotations.NotNull;

import java.io.OutputStream;
import java.util.logging.Level;
import java.util.logging.Logger;

public final class LoggingOutputStream extends OutputStream implements AutoCloseable {
    private Logger logger;
    private Level level;
    private char separator;
    private int count;
    private byte[] bytes;

    @Override
    public void write(int b) {
        if (separator != '\0' && b == separator) {
            flush();
        } else {
            if (count + 1 == bytes.length)
                flush();

            bytes[count] = (byte) b;
            count++;
        }
    }

    @Override
    public void write(@NotNull byte[] b) {
        logger.log(level, new String(b));
    }


    @Override
    public void flush() {
        if (count > 0) {
            String str = new String(bytes, 0, bytes[count - 1] == '\r' ? count - 1 : count);
            count = 0;

            logger.log(level, str);
        }
    }

    @Override
    public void close() {
        flush();
    }

    public LoggingOutputStream(@NotNull Logger logger, Level level, char separator, int bufferSize) {
        this.logger = logger;
        this.level = level;
        this.separator = separator;

        bytes = new byte[bufferSize];
    }

    public LoggingOutputStream(@NotNull Logger logger, Level level) {
        this(logger, level, '\n', 1024);
    }

    public LoggingOutputStream(@NotNull Logger logger, Level level, char separator) {
        this(logger, level, separator, 1024);
    }

    public LoggingOutputStream(@NotNull Logger logger, Level level, int bufferSize) {
        this(logger, level, '\n', bufferSize);
    }

    public int getCount() {
        return count;
    }

    @Override
    public String toString() {
        return "LoggingOutputStream for logger ${logger.name}";
    }
}

File: src/main/java/com/kenvix/utils/log/LoggingOutputStream.kt
//--------------------------------------------------
// Class LoggingOutputStream
//--------------------------------------------------
// Written by Kenvix <i@kenvix.com>
//--------------------------------------------------
package com.kenvix.utils.log

import com.kenvix.utils.lang.StringOutputStream
import org.slf4j.Logger
import java.io.OutputStream
import java.util.function.Consumer

@Suppress("MemberVisibilityCanBePrivate")
class LoggingOutputStream @JvmOverloads constructor(
    private val logger: Logger,
    separator: Char = '\n',
    bufferSize: Int = 1024,
    var onAccept: Consumer<String> = Consumer { s ->
        logger.info(s)
    }
) : OutputStream(), AutoCloseable, Consumer<String> {

    val count: Int
        get() = stream.count

    private val stream: StringOutputStream = StringOutputStream(this, separator, bufferSize)

    override fun write(b: Int) = stream.write(b)
    override fun write(b: ByteArray) = onAccept.accept(String(b))
    override fun flush() = stream.flush()
    override fun close() = flush()

    constructor(logger: Logger, bufferSize: Int) : this(logger, '\n', bufferSize)

    /**
     * Return current current buffer data as string
     * @return String data
     */
    override fun toString(): String {
        return stream.toString()
    }

    override fun accept(t: String) = onAccept.accept(t)
}